pkg/api/config.go

```diff
@@ -3,71 +3,228 @@ package api
-			}
-			if testConfig := test.OpenshiftAnsibleClusterTestConfiguration; testConfig != nil {
-				typeCount += 1
-				errs = append(errs, validateTargetCloud(test.As, testConfig.TargetCloud))
-			}
-			if testConfig := test.OpenshiftAnsibleSrcClusterTestConfiguration; testConfig != nil {
-				typeCount += 1
-				errs = append(errs, validateTargetCloud(test.As, testConfig.TargetCloud))
-			}
-			if testConfig := test.OpenshiftInstallerClusterTestConfiguration; testConfig != nil {
-				typeCount += 1
-				errs = append(errs, validateTargetCloud(test.As, testConfig.TargetCloud))
-			}
-			if testConfig := test.OpenshiftInstallerSmokeClusterTestConfiguration; testConfig != nil {
-				typeCount += 1
-				errs = append(errs, validateTargetCloud(test.As, testConfig.TargetCloud))
-			}
-			if typeCount == 0 {
-				// TODO remove when the migration is completed
-				if len(test.From) == 0 {
-					errs = append(errs, fmt.Errorf("test %q has no type", test.As))
-				}
-			} else if typeCount == 1 {
-				// TODO remove when the migration is completed
-				if len(test.From) > 0 {
-					errs = append(errs, fmt.Errorf("test %q specifies both `From` and a test type", test.As))
-				}
-			} else if typeCount > 1 {
-				errs = append(errs, fmt.Errorf("test %q has more than one type", test.As))
-			}
+	var validationErrors []error
+
+	validationErrors = append(validationErrors, validateReleaseBuildConfiguration(config))
+	validationErrors = append(validationErrors, validateBuildRootImageConfiguration("build_root", config.InputConfiguration.BuildRootImage))
+	validationErrors = append(validationErrors, validateTestStepConfiguration("tests", config.Tests))
+
+	if config.InputConfiguration.BaseImages != nil {
+		validationErrors = append(validationErrors, validateImageStreamTagReferenceMap("base_images", config.InputConfiguration.BaseImages))
+	}
+
+	if config.InputConfiguration.BaseRPMImages != nil {
+		validationErrors = append(validationErrors, validateImageStreamTagReferenceMap("base_rpm_images", config.InputConfiguration.BaseRPMImages))
+	}
+
+	if config.PromotionConfiguration != nil && config.InputConfiguration.ReleaseTagConfiguration != nil {
+		validationErrors = append(validationErrors, validatePromotionWithTagSpec(*config.PromotionConfiguration, *config.InputConfiguration.ReleaseTagConfiguration))
+	} else if config.PromotionConfiguration == nil && config.InputConfiguration.ReleaseTagConfiguration != nil {
+		validationErrors = append(validationErrors, validateReleaseTagConfiguration("tag_specification", *config.InputConfiguration.ReleaseTagConfiguration))
+	} else if config.PromotionConfiguration != nil && config.InputConfiguration.ReleaseTagConfiguration == nil {
+		validationErrors = append(validationErrors, validatePromotionConfiguration("promotion", *config.PromotionConfiguration))
+	}
+
+	return kerrors.NewAggregate(validationErrors)
+}
+
+func validatePromotionWithTagSpec(promotion PromotionConfiguration, tagSpec ReleaseTagConfiguration) error {
```

Comment:
So this is still wrong....

If we have both, both need to be validated. The release tag needs to be validated, full stop. The promotion needs to be validated only if the name or tag is specified